File: python/radar/targets.py
# SPDX-License-Identifier: Apache-2.0
# Copyright © 2025 Au-Zone Technologies. All Rights Reserved.

import rerun as rr
import zenoh
import sys
import asyncio
from argparse import ArgumentParser
from edgefirst.schemas import decode_pcd
from edgefirst.schemas.sensor_msgs import PointCloud2

def format_remote_endpoint(remote):
    if not remote:
        return None
    
    # Already in full format (tcp/IP:PORT)
    if remote.startswith("tcp/") and ":" in remote.split("/", 1)[1]:
        return remote
    
    # Just IP address (e.g., "10.10.41.100")
    if "/" not in remote and ":" not in remote:
        return f"tcp/{remote}:7447"
    
    # IP:PORT format (e.g., "10.10.41.100:7447")
    if ":" in remote and not remote.startswith("tcp/"):
        return f"tcp/{remote}"
    
    # tcp/IP format (e.g., "tcp/10.10.41.100")
    if remote.startswith("tcp/") and ":" not in remote.split("/", 1)[1]:
        return f"{remote}:7447"
    
    # Fallback: return as-is
    return remote

targets_msg = None


def targets_handler(msg):
    global targets_msg
    targets_msg = msg


async def targets_worker():
    global targets_msg
    while True:
        if targets_msg is None:
            await asyncio.sleep(0.001)
            continue

        try:
            pcd = PointCloud2.from_cdr(targets_msg.payload.to_bytes())
        except Exception as e:
            targets_msg = None
            print(f"Error processing radar targets: {e}", file=sys.stderr)
            continue

        targets_msg = None
        points = decode_pcd(pcd)
        pos = [[p.x, p.y, p.z] for p in points]
        rr.log("radar/targets", rr.Points3D(pos))


async def main_async(args):
    # Setup rerun
    args.memory_limit = 10
    rr.script_setup(args, "radar/targets")

    # Zenoh config
    config = zenoh.Config()
    config.insert_json5("scouting/multicast/interface", "'lo'")
    if args.remote:
        remote_endpoint = format_remote_endpoint(args.remote)
        config.insert_json5("mode", "'client'")
        config.insert_json5("connect", f'{{"endpoints": ["{remote_endpoint}"]}}')
    session = zenoh.open(config)

    session.declare_subscriber("rt/radar/targets", targets_handler)
    targets_task = asyncio.create_task(targets_worker())

    try:
        await asyncio.gather(targets_task)
    finally:
        targets_task.cancel()
        await asyncio.gather(targets_task, return_exceptions=True)



def main():
    parser = ArgumentParser(description="EdgeFirst Samples - Radar Targets")
    parser.add_argument(
        "-r",
        "--remote",
        type=str,
        default=None,
        help="Connect to remote endpoint (e.g., '10.10.41.100' or 'tcp/10.10.41.100:7447').",
    )
    rr.script_add_args(parser)
    args = parser.parse_args()

    try:
        asyncio.run(main_async(args))
    except KeyboardInterrupt:
        sys.exit(0)


if __name__ == "__main__":
    main()

File: src/radar/targets.rs
use clap::Parser;
use edgefirst_schemas::{decode_pcd, sensor_msgs::PointCloud2};
use rerun::{Points3D, Position3D};
use std::{error::Error, path::PathBuf, time::Instant};
use zenoh::Config;
#[derive(Parser, Debug, Clone)]
struct Args {
    /// Time in seconds to run command before exiting.
    #[arg(short, long)]
    pub timeout: Option<u64>,

    /// Connect to a Zenoh router rather than peer mode.
    #[arg(short, long)]
    connect: Option<String>,

    /// Rerun file
    #[arg(short, long)]
    rerun: Option<PathBuf>,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let args = Args::parse();

    let rec = rerun::RecordingStreamBuilder::new("radar/targets Example")
        .save(args.rerun.unwrap_or("radar-targets.rrd".into()))?;

    // Create the default Zenoh configuration and if the connect argument is
    // provided set the mode to client and add the target to the endpoints.
    let mut config = Config::default();
    if let Some(connect) = args.connect {
        let post_connect = format!("['{connect}']");
        config.insert_json5("mode", "'client'").unwrap();
        config
            .insert_json5("connect/endpoints", &post_connect)
            .unwrap();
    }
    let session = zenoh::open(config).await.unwrap();

    // Create a subscriber for "rt/radar/targets"
    let subscriber = session
        .declare_subscriber("rt/radar/targets")
        .await
        .unwrap();

    let start = Instant::now();

    while let Ok(msg) = subscriber.recv() {
        if let Some(timeout) = args.timeout {
            if start.elapsed().as_secs() >= timeout {
                break;
            }
        }
        let pcd: PointCloud2 = cdr::deserialize(&msg.payload().to_bytes())?;
        let points = decode_pcd(&pcd);
        let min_x = points.iter().map(|p| p.x).fold(f64::INFINITY, f64::min);
        let max_x = points.iter().map(|p| p.x).fold(f64::NEG_INFINITY, f64::max);

        let min_y = points.iter().map(|p| p.y).fold(f64::INFINITY, f64::min);
        let max_y = points.iter().map(|p| p.y).fold(f64::NEG_INFINITY, f64::max);

        let min_z = points.iter().map(|p| p.z).fold(f64::INFINITY, f64::min);
        let max_z = points.iter().map(|p| p.z).fold(f64::NEG_INFINITY, f64::max);

        let min_speed = points
            .iter()
            .map(|p| *p.fields.get("speed").unwrap())
            .fold(f64::INFINITY, f64::min);
        let max_speed = points
            .iter()
            .map(|p| *p.fields.get("speed").unwrap())
            .fold(f64::NEG_INFINITY, f64::max);

        println!(
            "Recieved {} radar points. Values: x: [{min_x:.2}, {max_x:.2}]\ty: [{min_y:.2}, {max_y:.2}]\tz: [{min_z:.2}, {max_z:.2}]\tspeed: [{min_speed:.2}, {max_speed:.2}]",
            points.len(),
        );

        let rr_points = Points3D::new(
            points
                .iter()
                .map(|p| Position3D::new(p.x as f32, p.y as f32, p.z as f32)),
        );
        let _ = rec.log("radar/targets", &rr_points);
    }

    Ok(())
}
